torch/csrc/autograd/profiler_python.cpp

```diff
@@ -0,0 +1,724 @@
+#include <torch/csrc/autograd/profiler_python.h>
+
+#include <iostream>
+#include <limits>
+#include <memory>
+
+#if CAFFE2_HAVE_RE2
+#include <re2/re2.h>
+#else
+#include <regex>
+#endif
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <Python.h>
+#include <frameobject.h>
+
+#include <c10/macros/Macros.h>
+#include <c10/util/flat_hash_map.h>
+#include <torch/csrc/autograd/profiler_kineto.h>
+#include <torch/csrc/utils/python_strings.h>
+#include <torch/csrc/utils/pybind.h>
+
+namespace py = pybind11;
+
+
+namespace torch { namespace autograd { namespace profiler { namespace python_tracer {
+namespace {
+
+// ============================================================================
+// == Core data types =========================================================
+// ============================================================================
+
+// PyObject that allows different threads to record events without colliding.
+// It is passed as the second argument when enabling tracing via
+// `PyEval_SetProfile`.
+struct TraceContext {
+    PyObject_HEAD
+
+    // It is wasteful to store an entire PyThreadState*. Instead, we map thread
+    // ids down to a compact space that we can store in a single byte.
+    uint8_t thread_id_;
+    PyThreadState* thread_state_;
+
+    // Likewise, int64_t is more precision than we need. By tracking when the
+    // profiler starts we can store "time since profile begin" which can fit
+    // into less space.
+    int64_t initial_us_;
+
+    // TODO:
+    //   Wall time is actually fairly expensive to compute. Empirically, it
+    //   takes ~600 ns to call `now()`. This puts a hard lower bound on the
+    //   overhead of the tracer. If we collected wall time less frequently, and
+    //   used TSC (e.g. through __rdtsc) to interpolate it should be possible
```

Comment:
It looks like https://github.com/google/benchmark/blob/main/src/cycleclock.h is a good candidate. TSC is a pain because:
1) It's so platform and chip dependent. (As evidenced by the fact that `cycleclock.h` is literally 20% conditional directives...)
2) TSC to time conversions are not well defined and have to be determined empirically. (A subset of chips let you query it.)

That said, I think it's worth doing. Not just because of the overhead reduction, but also because you're waaaayyyy less likely to collide on TSC than on microseconds and correctly ordering events when interleaving them is going to be important as we scale to a many profiler model.